### PR TITLE
Allow MessageBuilder, MessageHeader and ErrorHeader to be subclassed

### PR DESCRIPTION
**Pull request by aliddeke**
Port over issue #282 changes to C++.

### CODE REVIEW

File: core/src/main/cpp/msg/MessageBuilder.h [hunk 1]
```diff
@@ -62,6 +63,13 @@ class MessageBuilder : public std::enable_shared_from_this<MessageBuilder>
      */
     static int64_t decrementMessageId(const int64_t messageId);
 
+    /**
+     * <p>Minimal constructor that sets the MSL context</p>
+     *
+     * @param ctx MSL context.
+     */
+    MessageBuilder(std::shared_ptr<util::MslContext> ctx) : ctx_(ctx) {}
```
Comment (wmiaw) [inline core/src/main/cpp/msg/MessageBuilder.h]: Shouldn't this be `protected`?

File: core/src/main/cpp/msg/MessageBuilder.h [hunk 2]
```diff
@@ -519,9 +434,72 @@ class MessageBuilder : public std::enable_shared_from_this<MessageBuilder>
...
+            int64_t messageId,
+            std::shared_ptr<msg::MessageCapabilities> capabilities,
+            std::shared_ptr<tokens::MasterToken> masterToken,
+            std::shared_ptr<tokens::UserIdToken> userIdToken,
+            std::set<std::shared_ptr<tokens::ServiceToken>> serviceTokens,
+            std::shared_ptr<tokens::MasterToken> peerMasterToken,
+            std::shared_ptr<tokens::UserIdToken> peerUserIdToken,
+            std::set<std::shared_ptr<tokens::ServiceToken>> peerServiceTokens,
+            std::shared_ptr<keyx::KeyExchangeFactory::KeyExchangeData> keyExchangeData);
+    /**
```
Comment (wmiaw) [inline core/src/main/cpp/msg/MessageBuilder.h]: Missing newline.

File: core/src/main/cpp/msg/MessageBuilder.h [hunk 3]
```diff
@@ -519,9 +434,72 @@ class MessageBuilder : public std::enable_shared_from_this<MessageBuilder>
...
+            std::shared_ptr<util::MslContext> ctx,
+            std::shared_ptr<entityauth::EntityAuthenticationData> entityAuthData,
+            std::shared_ptr<tokens::MasterToken> masterToken,
+            std::shared_ptr<MessageHeader::HeaderData> headerData,
+            std::shared_ptr<MessageHeader::HeaderPeerData> peerData);
+
+protected:
     /** MSL context. */
-    std::shared_ptr<util::MslContext> ctx_;
+    const std::shared_ptr<util::MslContext> ctx_;
```
Comment (wmiaw) [inline core/src/main/cpp/msg/MessageBuilder.h]: I'm not sure `const` on a `shared_ptr` matters. @padolph

File: core/src/main/cpp/msg/MessageHeader.h [hunk 4]
```diff
@@ -1,3 +1,19 @@
+/**
+ * Copyright (c) 2016-2018 Netflix, Inc.  All rights reserved.
```
Comment (wmiaw) [inline core/src/main/cpp/msg/MessageHeader.h]: If the copyright text was missing, we should make sure the start year is correct.

**Comment by aliddeke** [comment]
@wmiaw  Yeah, I noticed that there are tabs/spaces mixed within the code before I started adding my in my changes.  My editor prefers spaces, but I tried to match the local file where it made sense.  Some files I noticed had almost all spaces, others had all tabs, while others had spaces+tabs.

File: core/src/main/cpp/msg/MessageBuilder.h (hunk 1 above)
Comment (aliddeke) [reply core/src/main/cpp/msg/MessageBuilder.h]: I agree.  I'll make the change.

File: core/src/main/cpp/msg/MessageHeader.h (hunk 4 above)
Comment (aliddeke) [reply core/src/main/cpp/msg/MessageHeader.h]: @padolph was it 2016 when you first created this file?

Comment (wmiaw) [reply core/src/main/cpp/msg/MessageHeader.h]: Checked file history. 2016 is the correct start year.

File: core/src/main/cpp/msg/MessageBuilder.h (hunk 2 above)
Comment (aliddeke) [reply core/src/main/cpp/msg/MessageBuilder.h]: fixed.

File: core/src/main/cpp/msg/MessageBuilder.h (hunk 3 above)
Comment (aliddeke) [reply core/src/main/cpp/msg/MessageBuilder.h]: @padolph convinced me of this, I'll take out the const.

**Pull request merged**